<h1 id="percentage-of-genome-covered-by-exons-not-in-pseudogenes">Percentage of genome covered by exons not in pseudogenes</h1>
<h2 id="preparing-files">Preparing files</h2>
<p>We will need the GFF file for mm10/GR38.75 and the txt file with the sequence information.</p>
<ul>
<li><em>Mus_musculus.GRCm38.75_chr1.gtf.gz</em><br />
</li>
<li><em>Mus_musculus.GRCm38_genome.txt</em></li>
</ul>
<p>Both files comes from <a href="https://github.com/vsbuffalo/bds-files/tree/master/chapter-09-working-with-range-data" class="uri">https://github.com/vsbuffalo/bds-files/tree/master/chapter-09-working-with-range-data</a> and, as the README file in this repository says, you can get the data following this instructions:</p>
<ul>
<li><p>The first file comes from <a href="http://ftp.ensembl.org/pub/release-75/gtf/mus_musculus/" class="uri">http://ftp.ensembl.org/pub/release-75/gtf/mus_musculus/</a> and then we have selected those registers of the chromosome 1:</p>
<pre><code>  # Ref.: https://github.com/vsbuffalo/bds-files/tree/master/chapter-09-working-with-range-data
  gzcat Mus_musculus.GRCm38.75.gtf.gz | egrep &quot;^(1\t|#)&quot; | gzip &gt; Mus_musculus.GRCm38.75_chr1.gtf.gf</code></pre></li>
<li><p>The second file comes from:</p>
<pre><code>  # Ref.: https://github.com/vsbuffalo/bds-files/tree/master/chapter-09-working-with-range-data
  curl ftp://ftp.ensembl.org/pub/release-75/fasta/mus_musculus/dna/Mus_musculus.GRCm38.75.dna_rm.toplevel.fa.gz \
  bioawk -c fastx &#39;{print $name&quot;\t&quot;length($seq)}&#39; &gt; Mus_musculus.GRCm38_genome.txt</code></pre></li>
</ul>
<h2 id="unzip-gff-files">Unzip GFF files</h2>
<p>First of all, we have to unzip our GFF file with the features of mm10:</p>
<pre><code>$ gunzip Mus_musculus.GRCm38.75_chr1.gtf.gz</code></pre>
<h2 id="extract-exons-ranges">Extract exons ranges</h2>
<p>Now, we will extract the exon ranges (we are not incorporating pseudogenes):</p>
<pre><code>$ bioawk -c gff &#39;$feature ~ /exon/ {print $seqname,$start,$end,$score,$filter,$9}&#39; Mus_musculus.GRCm38.75_chr1.gtf | grep -v &quot;pseudogene&quot; | cut -f1,2,3,4,5 | head
1   3102016 3102125 .   +
1   3213609 3216344 .   -
1   3205901 3207317 .   -
1   3213439 3215632 .   -
1   3206523 3207317 .   -
1   3670552 3671498 .   -
1   3421702 3421901 .   -
1   3214482 3216968 .   -
1   3466587 3466687 .   +
1   3513405 3513553 .   +

$ bioawk -c gff &#39;$feature ~ /exon/ {print $seqname,$start,$end,$score,$filter,$9}&#39; Mus_musculus.GRCm38.75_chr1.gtf | grep -v &quot;pseudogene&quot; | cut -f1,2,3,4,5 &gt; exons.bed</code></pre>
<h2 id="merge-exons-that-overlap">Merge exons that overlap</h2>
<p>We first have to sort the file:</p>
<pre><code>$ sort -k1,1 -k2,2n exons.bed &gt; exons-sorted.bed

$ bedtools merge -i exons-sorted.bed | head
1   3102016 3102125
1   3205901 3207317
1   3213439 3216968
1   3421702 3421901
1   3466587 3466687
1   3513405 3513553
1   3670552 3671498
1   3783876 3783933
1   4343507 4350091
1   4351910 4352081

$ bedtools merge -i exons-sorted.bed &gt; exons-merged.bed</code></pre>
<h2 id="compute-the-width-of-all-the-merged-exons">Compute the width of all the merged exons</h2>
<p>We can do that using <code>bioawk</code>:</p>
<pre><code>$ bioawk -c bed &#39;BEGIN{ s = 0}; {s += ($3-$2)}; END{ print &quot;width: &quot; s }&#39; exons-merged.bed
width: 5410233</code></pre>
<p>We can save it in a variable to use later:</p>
<pre><code>$ exons_width=$(bioawk -c bed &#39;BEGIN{ s = 0 }; {s += ($3-$2)}; END{ print s }&#39; exons-merged.bed)

$ echo $exons_width
5410233</code></pre>
<h2 id="width-of-chromosome-1-genome">Width of chromosome 1 genome</h2>
<p>We have the information in the <em>Mus_musculus.GRCm38_genome.txt</em> file:</p>
<pre><code>$ head Mus_musculus.GRCm38_genome.txt
1   195471971
10  130694993
11  122082543
12  120129022
13  120421639
14  124902244
15  104043685
16  98207768
17  94987271
18  90702639

$ cut -f2 Mus_musculus.GRCm38_genome.txt | head -n1
195471971</code></pre>
<p>Again, we can save it as a variable:</p>
<pre><code>$ chr1_width=$(cut -f2 Mus_musculus.GRCm38_genome.txt | head -n1)

$ echo $chr1_width
195471971</code></pre>
<h2 id="computing-the-proportion-of-the-genome-covered-by-exons-not-in-pseudogenes">Computing the proportion of the genome covered by exons not in pseudogenes</h2>
<pre><code>$ echo &quot;$exons_width/$chr1_width&quot; | bc -l
.02767779427568160143</code></pre>
<h1 id="percentage-of-the-whole-genome-covered-by-exons-not-in-pseudogenes">Percentage of the whole genome covered by exons not in pseudogenes</h1>
<h1 id="preparing-files-1">Preparing files</h1>
<p>We will need the GFF file for mm10/GR38.75 and the txt file with the sequence information.</p>
<ul>
<li><em>Mus_musculus.GRCm39.104.gtf.gz</em><br />
</li>
<li><em>Mus_musculus.GRCm39_genome.txt</em></li>
</ul>
<p>The first file comes from $ curl http://ftp.ensembl.org/pub/release-104/gtf/mus_musculus/Mus_musculus.GRCm39.104.gtf.gz<br />
&gt; Mus_musculus.GRCm39.104.gtf.gz</p>
<p>The second file comes from:</p>
<pre><code>$ curl http://ftp.ensembl.org/pub/release-104/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna_sm.toplevel.fa.gz | bioawk -c fastx &#39;{print $name&quot;\t&quot;length($seq)}&#39; &gt; Mus_musculus.GRCm39_genome.txt  </code></pre>
<h2 id="extracting-exons-ranges">Extracting exons ranges</h2>
<p>We select all exons not belonging to pseudogenes:</p>
<pre><code>$ bioawk -c gff &#39;$feature ~ /exon/ {print $seqname,$start,$end,$score,$filter,$9}&#39; Mus_musculus.GRCm39.104.gtf | grep -v &quot;pseudogene&quot; | cut -f1,2,3,4,5 | head -n2
1   150956201   150958296   .   +
1   108344807   108347562   .   +</code></pre>
<p>Note that now we are working in the <strong>mm_39</strong> genome version.</p>
<h2 id="sorting-and-mergin-exons-ranges">Sorting and mergin exons ranges</h2>
<p>First, we sort by chromosome and by start position:</p>
<pre><code>$ sort -k1,1 -k2,2n exons-genome.bed &gt; exons-genome-sorted.bed</code></pre>
<p>Then, we can merge the exons ranges:</p>
<pre><code>$ bedtools merge -i exons-genome-sorted.bed &gt; exons-genome-merged.bed</code></pre>
<h2 id="filtering-for-chromosomes">Filtering for chromosomes</h2>
<p>We select only those registers about chromosome, not scaffolds or other sequences:</p>
<pre><code>$ bioawk &#39;$1 ~ &quot;^[1-9XYM]&quot; {print $0}&#39; exons-genome-merged.bed &gt; exons-genome-merged-chr.bed</code></pre>
<p>Now weâ€™ll do the same but for sequences lengths:</p>
<pre><code>$ grep -E &quot;^[1-9XYM]T*[^A-Za-z]&quot; Mus_musculus.GRCm39_genome.txt &gt; mm_genome-chr.txt</code></pre>
<h2 id="computing-the-total-width-of-exons-not-in-a-pseudogene-for-the-whole-genome">Computing the total width of exons not in a pseudogene for the whole genome</h2>
<pre><code>$ bioawk -c bed &#39;BEGIN{ s = 0 }; {s += ($end-$start)}; END{ print &quot;width: &quot; s}&#39; exons-genome-merged-chr.bed
width: 117833254</code></pre>
<p>We can save it in a variable to use later:</p>
<pre><code>$ exons_width=$(bioawk -c bed &#39;BEGIN{ s = 0 }; {s += ($end-$start)}; END{ print s}&#39; exons-genome-merged-chr.bed)</code></pre>
<h2 id="computing-the-total-width-for-the-whole-genome">Computing the total width for the whole genome</h2>
<pre><code>$ awk &#39;BEGIN{s = 0 }; {s += $2}; END{print &quot;width: &quot; s}&#39; mm_genome-chr.txt
width: 2.72343e+09</code></pre>
<p>Again, we can save it in a variable for later use:</p>
<pre><code>$ genome_width=$(awk &#39;BEGIN{s = 0 }; {s += $2}; END{print s}&#39; mm_genome-chr.txt)</code></pre>
<p>And we change the scientific notation format for another that <code>bc</code> can understant:</p>
<pre><code>$ echo &quot;$genome_width&quot; | sed -e &#39;s/e+/\*10\^/&#39;
2.72343*10^09

$ genome_width=$(echo &quot;$genome_width&quot; | sed -e &#39;s/e+/\*10\^/&#39;)

$ echo &quot;$genome_width&quot; | bc
2723430000.00000

$ genome_width=$(echo &quot;$genome_width&quot; | bc)</code></pre>
<h2 id="compute-the-result">Compute the result</h2>
<pre><code>$ echo &quot;$exons_width/$genome_width&quot; | bc -l
.04326648894959664834</code></pre>
